### ISSUE TITLE
Build fails if new artifact removes a class even with major version change

### ISSUE THREAD

**Issue by tecywiz121** [siom79/japicmp#190]
I'm using `<breakBuildBasedOnSemanticVersioning>true</breakBuildBasedOnSemanticVersioning>` to enforce proper semantic versioning in my maven builds.

I think removing a class is allowed if I bump the major version, but the plugin fails the build, and suggests I add `--ignore-missing-classes`.

**Comment by siom79** [comment]
Removing a class is not a compatible change. What if someone is already
using it?

tecywiz121 <notifications@github.com> schrieb am Do., 14. Dez. 2017, 07:09:

**Comment by tecywiz121** [comment]
Breaking changes are fair game if I bump the major version, no?

**Comment by twogee** [comment]
What about change of Java version? #168

**Comment by siom79** [comment]
Can you post the complete output?

twogee <notifications@github.com> schrieb am Do., 14. Dez. 2017, 08:34:

**Comment by tecywiz121** [comment]
Tried to create a minimal working example and realized that the class that was removed was in a dependency, not in the jar I'm trying to compare.

I'll finish up an example still if you think that'll help.

**Comment by siom79** [comment]
This explains the ignore missing classes hint you have seen. The tool does
not compare the transitive dependencies, too. Hence you don't have to
provide an example.

tecywiz121 <notifications@github.com> schrieb am Do., 14. Dez. 2017, 17:07:

**Comment by tecywiz121** [comment]
It still fails the build when this happens, which I think is incorrect.

Should the ignore missing classes option be enabled automatically when the major version changes? Or is it relatively safe to ignore missing classes all the time?

**Comment by siom79** [comment]
What relation does your code have to the missing class? Why does it compile
without this class?

tecywiz121 <notifications@github.com> schrieb am Do., 14. Dez. 2017, 17:50:

**Comment by tecywiz121** [comment]
We're doing some restructuring, and we moved the class into a different package. The new code has been updated to point at the new location.

**Comment by siom79** [comment]
Have you tried to configure japicmp using multiple old version artefacts?
This way you can for example create one report for all of your artefacts in
the parent pom file.

tecywiz121 <notifications@github.com> schrieb am Do., 14. Dez. 2017, 19:54:

**Comment by tecywiz121** [comment]
I'm not sure what you mean. What are multiple old version artifacts?

We don't actually have a single parent pom file. All of the projects build independently, which is why we care so much about semantic versioning.

**Comment by siom79** [comment]
Take a at this pom in the integration tests of japicmp:
https://github.com/siom79/japicmp/blob/master/japicmp-testbase/japicmp-test-maven-plugin/pom.xml

The execution multiple version shows how to specify more than one jar file.

If you have separate projects you could create a new project for japicmp
reporting and reference all the single projects there.

tecywiz121 <notifications@github.com> schrieb am Do., 14. Dez. 2017, 22:36:

**Comment by tecywiz121** [comment]
Ah, I think I see what you're saying. Do one giant japicmp for everything so that nothing is a transitive dependency.

That seems a little impractical since we'll have to include every dependency from every project (in case a dependency we use removes a class.) Is that correct, or am I misunderstanding?

**Comment by siom79** [comment]
Yes, you would need to include every dependency from every project that you
consider to be part of the API.

But somehow you must specify what is your API and what is not. Or do you
have a better idea?

tecywiz121 <notifications@github.com> schrieb am Do., 14. Dez. 2017, 22:57:

**Comment by tecywiz121** [comment]
I think it'd be much easier to automatically ignore missing classes if the major version changes. You sidestep the whole "what is part of the api" question.

**Comment by jjlharrison** [comment]
@siom79: I am seeing this problem too. I'm not sure I understand, is this behaviour by design? Why is a missing class not just marked as an incompatibility instead of an error?
